plugins/ltac/leminv.ml

```diff
@@ -256,15 +256,14 @@ let add_inversion_lemma_exn ~poly na com comsort bool tac =
 
 let lemInv id c =
   Proofview.Goal.enter begin fun gls ->
+  let env = Proofview.Goal.env gls in
   try
-    let clause = mk_clenv_from (pf_env gls) (project gls) (c, pf_get_type_of gls c) in
-    let clause = clenv_constrain_last_binding (EConstr.mkVar id) clause in
+    let clause = mk_clenv_from env (project gls) (c, pf_get_type_of gls c) in
+    let mv = List.last (clenv_arguments clause) in
+    let clause = clenv_instantiate mv clause (EConstr.mkVar id, Typing.type_of_variable env id) in
     Clenv.res_pf clause ~flags:(Unification.elim_flags ()) ~with_evars:false
   with
-    | NoSuchBinding ->
-        user_err
-          (hov 0 (pr_econstr_env (pf_env gls) (project gls) c ++ spc () ++ str "does not refer to an inversion lemma."))
-    | UserError _ ->
+    | Failure _ | UserError _ ->
```

Comment:
Can't we still convert the Failure into a dedicated exception (so readers don't have to wonder if clenv_instantiate or some other call in the block can raise Failure)? and once we do that we could keep the dedicated error message.

Reply:
This is annoying because I'd have to catch the error in List.last just to reraise another one locally. A lot of noise for no good reason, I think. IMO we should just stop exporting these exception-raising code in our base libraries, List.last should return an option and that's it.

(Also I'm not even sure whether the user error is still live. I've left it there out of safety. The leminv code is only used in super specific cases, I'm not even sure how to trigger those things.)

Reply:
>A lot of noise

IMO only a little noise and makes the code moderately easier to understand

You can also add a CList.last_opt if you prefer

Reply:
I'd still rather not. The point of this PR is to make a step towards the removal of Clenv.Internal.refiner, not to 1. make the standard API more complete and 2. clean up the code of legacy tactics for unclear corner cases.

Reply:
That doesn't mean you should make the code worse

Reply:
It's not worse, it's using the exceptions in the very same way. The try-with block should be removed but this is orthogonal to this PR.

Reply:
It is less clear due to using a less specific exception

Reply:
What exactly do you propose? If you catch the exception from List.last you still depend on some specific exception which may change in the future, and in addition you have to define a fresh one just for this piece of code. This looks pretty much like cluttering the code with noise to me. In any case, just push a commit atop of this PR since I don't really care neither in one direction nor the other.